### ISSUE TITLE
Race between Controller.Call and Call.Return

### ISSUE THREAD

**Issue by powerman** [golang/mock#369]
**Actual behavior** A clear and concise description of what the bug is.
```
==================
WARNING: DATA RACE
Read at 0x00c0002b0408 by goroutine 38:
  github.com/golang/mock/gomock.(*Controller).Call.func1()
      /home/powerman/go/pkg/mod/github.com/golang/mock@v1.3.1/gomock/call.go:401 +0x35c
  github.com/golang/mock/gomock.(*Controller).Call()
      /home/powerman/go/pkg/mod/github.com/golang/mock@v1.3.1/gomock/controller.go:217 +0xd7
  ...MY CODE WHICH CALLS SOME METHOD ON MOCK OBJECT...

Previous write at 0x00c0002b0408 by goroutine 24:
  github.com/golang/mock/gomock.(*Call).Return()
      /home/powerman/go/pkg/mod/github.com/golang/mock@v1.3.1/gomock/call.go:419 +0x45c
  ...MY TEST WHICH CALLS .EXPECT().….Return()...
==================
==================
WARNING: DATA RACE
Read at 0x00c0005940b8 by goroutine 38:
  github.com/golang/mock/gomock.(*Controller).Call()
      /home/powerman/go/pkg/mod/github.com/golang/mock@v1.3.1/gomock/controller.go:220 +0x10a
  ...MY CODE WHICH CALLS SOME METHOD ON MOCK OBJECT...

Previous write at 0x00c0005940b8 by goroutine 24:
  github.com/golang/mock/gomock.(*Call).Return()
      /home/powerman/go/pkg/mod/github.com/golang/mock@v1.3.1/gomock/call.go:419 +0x422
  ...MY TEST WHICH CALLS .EXPECT().….Return()...
==================
```

**Expected behavior** A clear and concise description of what you expected to happen.
I expect not to have race conditions in gomock. :smile: 

**To Reproduce** Steps to reproduce the behavior
From test, run some code which uses mocked object in parallel goroutine, then try to setup next EXPECT()ations in main test goroutine. Something like this:
```go
func Test(…) {
    …
    mockObj.EXPECT().….Return(…)
    go codeWhichUsesMockObj()
    mockObj.EXPECT().….Return(…)
    go codeWhichUsesMockObj()
    …
}
```
Actual code which triggers this issue is under NDA, but I believe information above should be enough to fix it.

**Additional Information**

-   gomock mode (reflect or source): source
-   gomock version or git ref: v1.3.1 (latest, just installed by `go get github.com/golang/mock/mockgen`)
-   golang version: go version go1.13.5 linux/amd64

**Triage Notes for the Maintainers**

**Comment by codyoss** [comment]
Hey @powerman, 

Would you mind providing a basic example of how to reproduce this. I was unable to trip the race detector with the information provided.

**Comment by codyoss** [comment]
Closing this for now due to lack of response. Please feel free to reopen with more info. Thanks.

**Comment by aaronjwood** [comment]
I am hitting this as well, I can provide an example of what triggers it. The test:

```
func TestTokenRefresher_RefreshForever(t *testing.T) {
	k8sClient := fake.NewSimpleClientset(&v1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name: "test",
		},
		Data: map[string][]byte{
			"token": []byte(refreshToken),
		},
	})

	refresher, err := jwt.NewTokenRefresher(k8sClient, "test", "")
	if err != nil {
		t.Fatal(err)
	}

	ctrl := gomock.NewController(t)
	defer ctrl.Finish()
	someServer := mockSome.NewMockMASKEDServer(ctrl)
	server := grpc.NewServer()
	some.RegisterMASKEDServer(server, someServer)
	lis, err := net.Listen("tcp", "localhost:0")
	if err != nil {
		t.Fatal(err)
	}

	go server.Serve(lis)
	err = refresher.Connect(lis.Addr().String(), libgrpc.Config{TLS: false})
	if err != nil {
		t.Fatal(err)
	}

	accessToken := "masked"
	someServer.EXPECT().RefreshToken(gomock.Any(), &empty.Empty{}).Return(&some.RefreshTokenResponse{
		Token: []byte(accessToken),
	}, nil)
	ctx, cancel := context.WithCancel(context.Background())
	ch := make(chan string)
	go refresher.RefreshForever(ctx, ch)
	if <-ch != accessToken {
		t.Fatal("Wrong token string")
	}

	wrongToken := "masked"
	someServer.EXPECT().RefreshToken(gomock.Any(), &empty.Empty{}).Return(&some.RefreshTokenResponse{
		Token: []byte(wrongToken),
	}, nil)
	if <-ch == accessToken {
		t.Fatal("Should have failed on wrong token string")
	}

	someServer.EXPECT().RefreshToken(gomock.Any(), &empty.Empty{}).Return(nil, errors.New(""))
	someServer.EXPECT().RefreshToken(gomock.Any(), &empty.Empty{}).Return(&someServer.RefreshTokenResponse{
		Token: []byte(accessToken),
	}, nil)
	if <-ch != accessToken {
		t.Fatal("Wrong token string")
	}

	cancel()
	_, ok := <-ch
	if ok {
		t.Fatal("Channel should have been closed from cancelling")
	}

	server.Stop()
}
```

The meat:

```
func (j *TokenRefresher) RefreshForever(ctx context.Context, ch chan<- string) {
	for {
		var claims *Claims
		var accessTokenStr string
		var err error

	inner:
		for {
			select {
			case <-ctx.Done():
				log.Infof("Stopping refreshing of access token")
				close(ch)
				return
			default:
				log.Infof("Refreshing access token")
				claims, accessTokenStr, err = j.RefreshAccessToken(ctx)
				if err != nil {
					log.Errorf("Failed to refresh access token: %s", err.Error())
					retryDuration := 5 * time.Second
					log.Warnf("Retrying refreshing of the access token in %s", retryDuration)
					time.Sleep(retryDuration)
					continue inner
				}

				ch <- accessTokenStr
				break inner
			}
		}

		refreshDuration := time.Duration(claims.RefreshInterval) * time.Minute
		log.Infof("Access token successfully refreshed, refreshing again in %s", refreshDuration)
		time.Sleep(refreshDuration)
	}
}
```

The failures:

```
==================
 WARNING: DATA RACE
 Read at 0x00c0001a4fb8 by goroutine 57:
   github.com/golang/mock/gomock.(*Call).call()
       /root/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/call.go:408 +0x366
   github.com/golang/mock/gomock.(*Controller).Call.func1()
       /root/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/controller.go:212 +0x2fc
   github.com/golang/mock/gomock.(*Controller).Call()
       /root/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/controller.go:217 +0xdc
   MASKED.(*MockMASKEDServer).RefreshToken()
       /root/go/pkg/mod/MASKED/mock.go:125 +0x19d
   MASKED._MASKED_RefreshToken_Handler()
       /root/go/pkg/mod/MASKED/lifecycle.pb.go:538 +0x306
   google.golang.org/grpc.(*Server).processUnaryRPC()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:1082 +0xa05
   google.golang.org/grpc.(*Server).handleStream()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:1405 +0x138b
   google.golang.org/grpc.(*Server).serveStreams.func1.1()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:746 +0xdc
 Previous write at 0x00c0001a4fb8 by goroutine 40:
   github.com/golang/mock/gomock.(*Call).addAction()
       /root/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/call.go:426 +0x4a5
   github.com/golang/mock/gomock.(*Call).Return()
       /root/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/call.go:191 +0x36c
   MASKEDjwt_test.TestTokenRefresher_RefreshForever()
       MASKED/refresher_test.go:184 +0xf2b
   testing.tRunner()
       /usr/local/go/src/testing/testing.go:1123 +0x202
 Goroutine 57 (running) created at:
   google.golang.org/grpc.(*Server).serveStreams.func1()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:744 +0xb8
   google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/internal/transport/http2_server.go:447 +0x16e5
   google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/internal/transport/http2_server.go:488 +0x3e4
   google.golang.org/grpc.(*Server).serveStreams()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:742 +0x1c1
   google.golang.org/grpc.(*Server).handleRawConn.func1()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:703 +0x4c
 Goroutine 40 (running) created at:
   testing.(*T).Run()
       /usr/local/go/src/testing/testing.go:1168 +0x5bb
   testing.runTests.func1()
       /usr/local/go/src/testing/testing.go:1439 +0xa6
   testing.tRunner()
       /usr/local/go/src/testing/testing.go:1123 +0x202
   testing.runTests()
       /usr/local/go/src/testing/testing.go:1437 +0x612
   testing.(*M).Run()
       /usr/local/go/src/testing/testing.go:1345 +0x3b3
   main.main()
       _testmain.go:55 +0x236
 ==================
 ==================
 WARNING: DATA RACE
 Read at 0x00c0000e4b38 by goroutine 57:
   github.com/golang/mock/gomock.(*Controller).Call()
       /root/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/controller.go:220 +0x116
   MASKED.(*MockMASKEDServer).RefreshToken()
       /root/go/pkg/mod/MASKED/mock.go:125 +0x19d
   MASKED._MASKED_RefreshToken_Handler()
       /root/go/pkg/mod/MASKED/lifecycle.pb.go:538 +0x306
   google.golang.org/grpc.(*Server).processUnaryRPC()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:1082 +0xa05
   google.golang.org/grpc.(*Server).handleStream()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:1405 +0x138b
   google.golang.org/grpc.(*Server).serveStreams.func1.1()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:746 +0xdc
 Previous write at 0x00c0000e4b38 by goroutine 40:
   github.com/golang/mock/gomock.(*Call).addAction()
       /root/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/call.go:426 +0x46b
   github.com/golang/mock/gomock.(*Call).Return()
       /root/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/call.go:191 +0x36c
   MASKEDjwt_test.TestTokenRefresher_RefreshForever()
       MASKED/refresher_test.go:184 +0xf2b
   testing.tRunner()
       /usr/local/go/src/testing/testing.go:1123 +0x202
 Goroutine 57 (running) created at:
   google.golang.org/grpc.(*Server).serveStreams.func1()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:744 +0xb8
   google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/internal/transport/http2_server.go:447 +0x16e5
   google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/internal/transport/http2_server.go:488 +0x3e4
   google.golang.org/grpc.(*Server).serveStreams()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:742 +0x1c1
   google.golang.org/grpc.(*Server).handleRawConn.func1()
       /root/go/pkg/mod/google.golang.org/grpc@v1.28.1/server.go:703 +0x4c
 Goroutine 40 (running) created at:
   testing.(*T).Run()
       /usr/local/go/src/testing/testing.go:1168 +0x5bb
   testing.runTests.func1()
       /usr/local/go/src/testing/testing.go:1439 +0xa6
   testing.tRunner()
       /usr/local/go/src/testing/testing.go:1123 +0x202
   testing.runTests()
       /usr/local/go/src/testing/testing.go:1437 +0x612
   testing.(*M).Run()
       /usr/local/go/src/testing/testing.go:1345 +0x3b3
   main.main()
       _testmain.go:55 +0x236
 ==================
```

**Comment by codyoss** [comment]
@aaronjwood Would you mind opening a new issue with a simplified reproducer?
